src/Raven.Server/Documents/Indexes/Index.cs

```diff
@@ -936,20 +936,26 @@ public bool DoIndexingWork(IndexingStatsScope stats, CancellationToken cancellat
                         CommitStats commitStats;
                         tx.InnerTransaction.LowLevelTransaction.RetrieveCommitStats(out commitStats);
 
+                        tx.InnerTransaction.LowLevelTransaction.AfterCommitWhenNewReadTransactionsPrevented += () =>
+                        {
+                            if (writeOperation.IsValueCreated)
+                            {
+                                using (stats.For(IndexingOperation.Lucene.RecreateSearcher))
+                                {
+                                    // we need to recreate it after transaction commit to prevent it from seeing uncommitted changes
+                                    // also we need this to be called when new read transaction are prevented in order to ensure
+                                    // that queries won't get the searcher having 'old' state but see 'new' changes committed here
+                                    // e.g. the old searcher could have a segment file in its in-memory state which has been removed in this tx
+                                    IndexPersistence.RecreateSearcher(tx.InnerTransaction);
```

Comment:
No, this is cheap operation because we changed things that the actual searcher recreation is performed when a first query attempts to get the searcher (2933eadce5ffa313a7), instead of creating it after every indexing run. Underneath it's a lazy value:
https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexSearcherHolder.cs#L89